src/dmd/escape.d

```diff
@@ -590,7 +590,7 @@ bool checkAssignEscape(Scope* sc, Expression e, bool gag)
     // Try to infer 'scope' for va if in a function not marked @system
     bool inferScope = false;
     if (va && sc.func && sc.func.type && sc.func.type.ty == Tfunction)
-        inferScope = (cast(TypeFunction)sc.func.type).trust != TRUST.system;
+        inferScope = (cast(TypeFunction)sc.func.type).trust == TRUST.safe;
```

Comment:
The issue is about invalid error messages in normal functions, so it might be sensible to check whether the safety is inferred for the current function.

Reply:
That's what I'm trying now, but it means this doesn't compile:
```
auto ad(scope int* x)
{
    auto y = x;
    return y; // error: returning scope variable y
}
```
I'm thinking maybe scope inference should be done always, but then checkReturnEscape/checkAssignEscape should not error when returning an inferred scope variable unless it's a `@safe` function. Or maybe it shouldn't at all, should this compile?
```D
int* f(scope int* x) @system 
{
    return x;
}
```
It currently doesn't with -dip1000, which arguably should be fixed too, unless you don't count that as "scope escape analysis" but simply the trivial case of directly returning a scope variable. After all, this isn't allowed, dip1000 or not:
```D
int* f(int x) @system 
{
    return &x; // Error: returning `& x` escapes a reference to local variable `x`
}
```

Reply:
The first and second examples are fine because they might be valid depending on their usage ("scope semantics must be manually enforced"). But the third example could be rejected because it's known to always produce a stale pointer.

Reply:
I now removed the `inferScope` variable so inference happens always whether it's `@system`, default, interred, `@trusted` or `@safe`. Instead I added a `sc.func.setUnsafe()` check so it doesn't complain about returning `scope` variables in non-`@safe` code anymore.